### ISSUE TITLE
Implicit layers should be after `ppEnabledLayerNames`

### ISSUE THREAD

**Issue by krOoze** [KhronosGroup/Vulkan-Docs#1330]
It feels implicit layers (such as overlays) should be after `ppEnabledLayerNames` (i.e. closest to the driver). Or perhaps both places should be allowed based on some setting. But spec says otherwise:

> Implicitly enabled layers are loaded before explicitly enabled layers, such that implicitly enabled layers are closer to the application, and explicitly enabled layers are closer to the driver.

This is a problem e.g. in that the implicit layer would run even before the Validation Layers, possibly crashing before a diagnostic on error is issued.

**Comment by sfricke-samsung** [comment]
You don't have to use the Validation Layers as explicit, you can have it first as an implicit layer already if you want to validate everything prior to any other layers touching the calls

**Comment by krOoze** [comment]
Per the [LoaderAndLayerInterface.md](https://github.com/KhronosGroup/Vulkan-Loader/blob/master/loader/LoaderAndLayerInterface.md#overall-layer-ordering) `VK_INSTANCE_LAYERS` env layers are also enabled only after implicit layers, so that makes no difference.

Either way it is only an example.  Implicit layers being closest to the driver sounds generally like a more sane default to me. Mostly they are overlays and whatnot.

**Comment by DadSchoorse** [comment]
> This is a problem e.g. in that the implicit layer would run even before the Validation Layers, possibly crashing before a diagnostic on error is issued.

>  Implicit layers being closest to the driver sounds generally like a more sane default to me. Mostly they are overlays and whatnot.

When writing an implicit layer the current load order is very useful. If implicit layer were loaded closer to the driver than explicit layers the validation layers would not report errors caused by implicit layers. Of course it would be possible to work around this, but it would definitely be annoying.
 I don't really see a reason to have implicit layers enabled during normal vulkan application development on the other hand, so I'm not sure in what situation what you descripted is actually an issue. But maybe I'm missing something.
Anyway please leave the load order as is, or at least offer the current behavior as an option.

**Comment by krOoze** [comment]
It is a moral hazard to externalize development inconveniences to others. Not that I would even consider it one in this case, as you can force any layer order you wish during testing phase, which sounds like a routine problem to me. Implicit layers differ from explict ones only in metadata, not in the way they should be tested.

> I don't really see a reason to have implicit layers enabled during normal vulkan application development on the other hand, so I'm not sure in what situation what you descripted is actually an issue.

For example `VK_LAYER_NV_optimus`. For one, speaking about inconveniences, regular app developer should not be forced into actively disabling implicit layers. Besides this layer looks like part of the driver. Where should it be its natural place? Closest to your driver for which it is made, or closest to your app where its output goes through all your custom layers?

> or at least offer the current behavior as an option.

The more I think about it, the less I can imagine a hypothetical implicit layer for which the current order makes sense.

Only exception here feels to be `VK_INSTANCE_LAYERS`, which takes an explicit layers and turns them into implicit ones. Certainly, keep the order here, or make it an option.

**Comment by DadSchoorse** [comment]
> For example `VK_LAYER_NV_optimus`. For one, speaking about inconveniences, regular app developer should not be forced into actively disabling implicit layers. Besides this layer looks like part of the driver. Where should it be its natural place? Closest to your driver for which it is made, or closest to your app where its output goes through all your custom layers?

I agree, the load order you propose makes sense for this case.

> The more I think about it, the less I can imagine a hypothetical implicit layer for which the current order makes sense.

One example off the top of my head would be a layer with the purpose to fix one specific application that produces invalid api calls. If it would not be at the position closest to the application other layers could crash due to the invalid behavior.

**Comment by lenny-lunarg** [comment]
Honestly I'm not convinced that any particular order would be best. There can also be issues when a layer is in between a driver and other layers, as one layer may not actually see what the driver is outputting. The other reason I would be highly skeptical of making a change like this is that it will matter for pretty much every Vulkan app out there when using common tools like the validation layer, steam overlay/fossilize, and renderdoc. I would want to make sure we really understand the implication before doing something like that, and I doubt that the benefits would warrant such a wide-reaching change.

In my opinion, there are definitely issues with the way implicit layers are handled, but I think it makes more sense to address that by allowing the app more control over what goes on with their layers. That's a whole separate issue and I'm not thrilled at the prospect of changing the (default) ordering.

**Comment by TomOlson** [comment]
The working group discussed this - we agree with @lenny-lunarg that changing the default behavior would likely cause a lot of collateral damage in the ecosystem. Letting people opt in to other behaviors is thinkable and might be interesting to explore. It's not a high priority for us at the moment, however.
